Optimizing memory and adding a star

wardog21 opened the pull request.

wardog21 committed:
start optimizing memory by not comparing against colors and not creating tons of ball objects:
-global variables
-class Ball

wardog21 committed:
adding functional but not commented dirtyCode.py

wardog21 committed:
Merge branch 'main' of https://github.com/stealka/Experimenta_Logikspiel into optimizing-memory-and-adding-a-star

wardog21 committed:
-extending Game class
-minor bug fixes
-adding State class
-adding a-star algo

wardog21 committed:
-cleaning up
-adding comments where to add doxygen comments

wardog21 committed:
displaying solution after solving

wardog21 commented:
still have to add comments before merging

wardog21 committed:
adding solution

wardog21 committed:
Update README.md

adding game rules

wardog21 committed:
Merge branch 'optimizing-memory-and-adding-a-star' of https://github.com/stealka/Experimenta_Logikspiel into optimizing-memory-and-adding-a-star

stealka commented on main.py:
```diff
@@ -55,23 +65,36 @@ class Ball:
 
-        param[in]   coord   Playfield Array with circle properties
-                            (see COORDS_x[])
+        param[in]   id   Playfield Array id with circle properties
+                            (see COORDS[])
```
Description lineup doesn't match throughout the lines (optical/authistical)

wardog21 replied:
fixed

stealka commented on main.py:
```diff
@@ -32,11 +35,18 @@
+            (0,3),                      (4,3)]
+
+# Start and goal constellation - Level 1
+START = [1, 2, 1, 0, 0, 0, 2, 1, 0, 2, 1, 2]
+GOAL =  [2, 1, 2, 0, 0, 0, 1, 2, 0, 1, 2, 1]
```
[SUGGESTION] Substitute numbers with Macros of, ex. P1 (Player 1), P2 (Player 2), N (Playfield), to make the subsequent code more readable/understandable.

wardog21 replied:
applied as suggested

stealka commented on main.py:
```diff
@@ -32,11 +35,18 @@
+            (0,3),                      (4,3)]
+
+# Start and goal constellation - Level 1
+START = [1, 2, 1, 0, 0, 0, 2, 1, 0, 2, 1, 2]
+GOAL =  [2, 1, 2, 0, 0, 0, 1, 2, 0, 1, 2, 1]
```
```suggestion

START = [1,          2, 
         1, 0, 0, 0, 2,
         1,    0,    2, 
         1,          2]
 
GOAL =  [2,          1,
         2, 0, 0, 0, 1,
         2,    0,    1,
         2,          1]
```
Or variant with macros:
```suggestion
START = [P1,             P2, 
         P1,  N,  N,  N, P2,
         P1,      N,     P2, 
         P1,             P2]
 
GOAL =  [P2,             P1,
         P2,  N,  N,  N, P1,
         P2,      N,     P1,
         P2,             P1]

```

wardog21 replied:
applied as suggested

stealka commented on main.py:
```diff
@@ -32,11 +35,18 @@
 
+# Colours - Playfield, Player1, Player2
+COLOURS = [WHITE, BLACK, RED]
+
```
```suggestion
# Numbercode for ball colours
N  = 0    # Playfield
P1 = 1    # Player 1
P2 = 2    # Player 2
```

wardog21 replied:
applied as suggested

stealka commented on main.py:
```diff
@@ -164,6 +160,43 @@ def draw(self):
         img = g_font.render("Moves: " + str(self.counter), True, BLACK)
         g_screen.blit(img, (20, 20))
 
+    # add comments
+    def move(self, ball):
```
```suggestion
    def move(self, ball):
    """
    Function Ball.move(self, ball):
    Evaluating the imminent move depending on wether a ball is already selected and/or the status of selection the currently focused ball has.
    
    param[in]    ball    Instance of the currently observed ball
    """
```

wardog21 replied:
added comment

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+                self.move(ball)
                 break
 
+    # add comments
```
Todo

wardog21 replied:
added comment

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+            self.balls[i].selected = BALL_NOMARK
+            self.isBallSelected = False
+
+    # add comments
```
Todo

wardog21 replied:
added comment

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+            state.append(self.balls[i].colour)
+        return state
+
+# add comments
```
Todo

wardog21 replied:
added comment

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+        return state
+
+# add comments
+class State:
```
Class desctiption is missing

wardog21 replied:
added description

stealka replied:
Can't find it.
Can be shifted to a later merge.
```suggestion
class State:
"""
The class State: 
TODO[ADD] Description
"""
```

wardog21 replied:
added "TODO" as suggested

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+
+# add comments
+class State:
+    def __init__(self, id, inputState, parent, moves):
```
Todo: Function description

wardog21 replied:
added description

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+        self.stateValue = 0
+        self.calculateDistance()
+
+    def calculateDistance(self):
```
Todo: Function description

wardog21 replied:
added description

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+            self.distanceToGoal += 1
+        self.stateValue = self.distanceToStart + self.distanceToGoal
+
+    def updateParent(self, newParent, newMoves):
```
Todo: Function description

wardog21 replied:
added description

stealka commented on main.py:
```diff
@@ -176,30 +209,68 @@ def clicked(self, position):
+        self.distanceToStart = newMoves
+        self.calculateDistance()
+
+    def getNeighbours(self):
```
Todo: Function description

wardog21 replied:
added description

stealka commented on main.py:
```diff
@@ -226,6 +297,78 @@ def main():
     # Create a new game
     game = Game()
 
+    # add comments
```
Todo

wardog21 replied:
added comment

stealka commented on main.py:
```diff
@@ -226,6 +297,78 @@ def main():
 
+    # add comments
+    solving = False
+    # add comments
```
Todo

wardog21 replied:
added comments

stealka commented on main.py:
```diff
@@ -226,6 +297,78 @@ def main():
+                                displayedMove += str(COORDS[i])
+                                break
+                        displayedMove += " -> "
+                        for i in range(len(previousState)):
+                            if previousState[i] > 0 and currentState[i] == 0:
+                                displayedMove += str(COORDS[i])
+                                break
+                        print(displayedMove)
+                        previousState = currentState
+                        stateToBeCalculated = states[states[stateToBeCalculated].parent].id
+                    currentState = START
+                    displayedMove = ""
+                    for i in range(len(previousState)):
+                        if previousState[i] == 0 and currentState[i] > 0:
+                            displayedMove += str(COORDS[i])
+                            break
+                    displayedMove += " -> "
+                    for i in range(len(previousState)):
+                        if previousState[i] > 0 and currentState[i] == 0:
+                            displayedMove += str(COORDS[i])
+                            break
+                    print(displayedMove)
+                    break
+
+                stateInList = False
+                for state in states:
+                    if state.state == stateNeighbour:
+                        if state.distanceToStart > states[stateToBeCalculated].distanceToStart + 1:
+                            state.updateParent(stateToBeCalculated, states[stateToBeCalculated].distanceToStart + 1)
+                        stateInList = True
+                        break
+                if not stateInList:
+                    newId = len(states)
+                    states.append(State(newId, stateNeighbour, stateToBeCalculated, states[stateToBeCalculated].distanceToStart + 1))
+                    checkStates.append(newId)
+
+            if len(checkStates) == 0:
+                print("no solution found")
+                running = False
+
```
[SUGGESTION] Outsource code and depending class to another file.
Therefore main file would not get overloaded and we have the chance to use directory structures to maybe hide bruteforce solutions on the root directory.
(General directory structure could be a pending issue for the repository)

wardog21 replied:
Maybe we should create an issue for that.

stealka replied:
Issue is the way

stealka commented on solution.txt:
File could be moved from the root directory in a new "bruteforce" or "solutions" directory.
Just to make sure, that the game fun isn't directly ruined if new people get access to the repository.

wardog21 replied:
Maybe we should create an issue for that and talk about what we do want to present on the "first page" first.

stealka replied:
Issue is the way

stealka reviewed (commented):
Some suggestions and reminder to either add TODOs in Comments or process pending TODOs.

wardog21 committed:
-moved level to discrete file
-added comments and descriptions

stealka commented on main.py:
```diff
@@ -7,36 +7,36 @@
-LH_RED   = ( 255, 153, 153)
-GREEN    = (   0, 255,   0)
-WHITE    = ( 255, 255, 255)
+BLACK      = (   0,   0,   0)
```
Obsolete?
-> Since same define in levels.py

wardog21 replied:
moved colors to new file globals.py. otherwise there would have been circular imports at some point

stealka commented on main.py:
```diff
@@ -164,6 +150,51 @@ def draw(self):
+    def move(self, ball):
+        """
+        Function Game.move(self, ball):
+        Move function, that handles handles the action when the player clicks on 
```
Repetition corrected:
```suggestion
        Move function, that handles the action when the player clicks on 
```

wardog21 replied:
fixed

stealka commented on main.py:
```diff
@@ -176,30 +207,109 @@ def clicked(self, position):
+        This class defines the states relation by adding a parent and used moves to 
+        the state constellation. Also neighbours can be calculated.
+
+        parma[in]   id          state id
+        parma[in]   inputState  real state data
+        parma[in]   parent      state parent according to the shortest route
+        parma[in]   moves       moves needed to reach this state from START
```
Typo corrected:
```suggestion
        param[in]   id          state id
        param[in]   inputState  real state data
        param[in]   parent      state parent according to the shortest route
        param[in]   moves       moves needed to reach this state from START
```

wardog21 replied:
fixed

stealka commented on main.py:
```diff
@@ -176,30 +207,109 @@ def clicked(self, position):
+        If there has been found a shorter path to the current state,
+        update the parent and used moves.
+
+        parma[in]   newParent   state id
+        parma[in]   newMoves    real state data
```
Typo corrected:
```suggestion
        param[in]   newParent   state id
        param[in]   newMoves    real state data
```

wardog21 replied:
fixed

stealka commented on main.py:
```diff
@@ -164,6 +150,51 @@ def draw(self):
+
+        # Case(1): No ball is selected atm and the observed ball is not white.
+        #   -> Select it and set depending variables accordingly.
+        if not self.isBallSelected and ball.colour != 0:
```
Does using of the macro makes sense here?:
```suggestion
        if not self.isBallSelected and ball.colour != N:
```

wardog21 replied:
adapted

stealka commented on main.py:
```diff
@@ -164,6 +150,51 @@ def draw(self):
+            checkBalls.append(ball.id)
+            while len(checkBalls) > 0:
+                for selectableBall in self.balls[checkBalls[0]].neighbours:
+                    if (self.balls[selectableBall].colour == 0 and 
```
Does using the macro makes sense here?
```suggestion
                    if (self.balls[selectableBall].colour == N and 
```

wardog21 replied:
adapted

stealka commented on main.py:
```diff
@@ -164,6 +150,51 @@ def draw(self):
+        #      all balls and set depending variables accordingly.
+        elif self.isBallSelected and ball.selected == BALL_MARK_S:
+            ball.colour = self.selectedBall.colour
+            self.selectedBall.colour = 0
```
Does using the macro makes sense here?
```suggestion
            self.selectedBall.colour = N
```

wardog21 replied:
adapted

stealka commented on main.py:
```diff
@@ -176,30 +207,109 @@ def clicked(self, position):
+        Function Game.setState(self, state):
+        Sets the state of the current game aka loading a game.
+
+        parma[in]   state   state to set / load
```
Corretected Typo:
```suggestion
        param[in]   state   state to set / load
```

wardog21 replied:
fixed

stealka reviewed (changes requested):
Fixed a few typos + 
Use of Colour Array in the code?

stealka commented on main.py:
```diff
@@ -176,30 +207,109 @@ def clicked(self, position):
+                self.distanceToGoal += 1
+                if self.state[index] == 0:
+                    noWhite = False
+        if noWhite:
+            self.distanceToGoal += 1
```
Does the indentation match here?

wardog21 replied:
Yes it does for the current way the distance function is used. An example is the distance comparions between START and GOAL. For the first and only level so far, all 8 balls have to be moved. But the distance have to be 9, since two balls can't just switch places. So there has to be an extra field used for all where the first ball can move to ignoring all the rules like "you can't jump over anyone". This is considered the extra move which is checked with noWhite.

stealka reviewed (commented):
One indentation issue opend

wardog21 committed:
adapt review findings

moved some globals to globals.py
fixed typos
added description "todos"
replaced empty fields with N instead of 0
disabled solving as default

stealka reviewed (approved):
All findings resolved.
Pull request approved.
